mantidproject/mantid
# Slice viewer improvements - extents and tick behavior changes

**Description of work.**

This PR addresses a few requests from users to improve the usability of slice viewer. 

1. Integer ticks/gridlines for hkl-frames
2. Ability to change extents/limits of slice
3. Fix uninteded bug with 4d MDHisto workspaces when loading/creating histo workspaces directly



**To test:**

Run this script to generate a 4d workspace with HKL frame and 1 general frame.

1. Ensure in ortho mode and non-ortho mode that integer-only tick marks are displayed when X and/or Y are in units of HKL. 
2. Zoom in to a region and make sure the limiting case with range smaller than 1.0 reverts to fractional ticks
3. Notice new min/max limit spin boxes in the dimension widget. Modify them by hand to ensure the frame extents get updated accordingly in ortho mode and non-ortho mode
5. Return to home to ensure everything gets set back to original
6. Zoom in and verify the zoomed extents match the display in the spin boxes

![image](https://user-images.githubusercontent.com/13754794/229773137-aa62dd7e-161d-4398-894b-c21bf56b0377.png)
![image](https://user-images.githubusercontent.com/13754794/229773233-9efa391f-253c-4950-a111-15a0495584b0.png)
![image](https://user-images.githubusercontent.com/13754794/229773288-1c27edac-63e0-47aa-a7db-e2405e51ccf0.png)

MDHisto script
```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

signal = np.linspace(0,1,14641)
error = np.sqrt(signal)

signal -= 0.5

signal[signal>0.25] = np.nan

ws = CreateMDHistoWorkspace(Dimensionality=4,
                            Extents='-3,3,-10,10,-5,5,0,1',
                            SignalInput=signal,
                            ErrorInput=error,
                            NumberOfBins='11,11,11,11',
                            Names='[H,0,0],[0,K,0],[0,0,L],values',
                            Units='r.l.u.,r.l.u.,r.l.u.,a.b.u.',
                            Frames='HKL,HKL,HKL,General Frame')

SetMDFrame('ws', MDFrame='HKL', Axes=[0,1,2])

AddSampleLog('ws', LogName='sample')
ws.getExperimentInfo(0).run().addProperty('W_MATRIX', list(np.eye(3).flatten()), True)

SetUB('ws', a=5, b=5, c=7, alpha=90, beta=90, gamma=120)
```

MDEvent script
```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws = CreateMDWorkspace(Dimensions='4', 
                       EventType='MDEvent',
                       Extents='-10,10,-10,10,-10,10,0,1',
                       Names='[H,0,0],[0,K,0],[0,0,L],values',
                       Units='r.l.u.,r.l.u.,r.l.u.,a.b.u.')

FakeMDEventData(ws, UniformParams='1000000')

SetMDFrame('ws', MDFrame='HKL', Axes=[0,1,2])

AddSampleLog('ws', LogName='sample')
ws.getExperimentInfo(0).run().addProperty('W_MATRIX', list(np.eye(3).flatten()), True)

SetUB('ws', a=5, b=5, c=7, alpha=90, beta=90, gamma=120)

CreatePeaksWorkspace(OutputWorkspace='peaks', OutputType='LeanElasticPeak')
```



*There is no associated issue.*





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
